src/hooks/useProducts.tsx

```diff
@@ -0,0 +1,54 @@
+import { useEffect, useState } from "react";
+import { ProductsAPI } from "../apis/products";
+import { CategoryOptionsKey, SortOptionsKey } from "../constants";
+import { Products } from "../types/products";
+import { isErrorResponse } from "../utils/typeGuard";
+
+type Props = React.Dispatch<React.SetStateAction<string>>;
+
+interface UseProductsReturn {
+  products: Products | null;
+  isLoading: boolean;
+  selectedCategory: CategoryOptionsKey;
+  setSelectedCategory: React.Dispatch<React.SetStateAction<CategoryOptionsKey>>;
+  selectedSortOption: SortOptionsKey;
+  setSelectedSortOption: React.Dispatch<React.SetStateAction<SortOptionsKey>>;
+}
+
+const useProducts = (setErrorMessage: Props): UseProductsReturn => {
+  const [products, setProducts] = useState<Products | null>(null);
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+  const [selectedCategory, setSelectedCategory] =
+    useState<CategoryOptionsKey>("전체");
+  const [selectedSortOption, setSelectedSortOption] =
+    useState<SortOptionsKey>("낮은 가격 순");
+
+  useEffect(() => {
+    (async () => {
+      setIsLoading(true);
+      const response = await ProductsAPI.get(
+        selectedCategory,
+        selectedSortOption
+      );
+      setIsLoading(false);
+
+      if (isErrorResponse(response)) {
+        setErrorMessage(response.error);
+        return;
+      }
+
+      setProducts(response as Products);
+    })();
+  }, [selectedCategory, selectedSortOption, setErrorMessage]);
+
+  return {
+    products,
+    isLoading,
+    selectedCategory,
+    setSelectedCategory,
+    selectedSortOption,
+    setSelectedSortOption,
+  };
```

Comment:
이렇게 namespace 묶어주면 밖에서 사용할 때 더 직관적이지 않을까요!?
```suggestion
  return {
    products: { value: products, loading: isLoading },
    category: { value: selectedCategory, change: setSelectedCategory },
    sortOption: { value: selectedSortOption, change: setSelectedSortOption   },
  };
```